File: LintCode/cpp/Maximum Product Subarray.cc
class Solution {
public:
    /**
     * @param nums: a vector of integers
     * @return: an integer
     */
    int maxProduct(vector<int>& nums) {
        int mini = 1;
        int maxi = 1;
        int ans = -INF;
        for (auto num: nums) {
            mini *= num;
            maxi *= num;
            
            auto tuple = {ans, maxi, mini, num};
            ans = accumulate(
                tuple.begin(), tuple.end(), -INF,
                [](const int& a, const int& b) {
                    return max(a, b);
                });
            
            int _maxi = max(max(maxi, mini), num);
            int _mini = min(min(maxi, mini), num);
            maxi = _maxi;
            mini = _mini;
        }
        return ans;
    }
private:
    static const int INF = 0x3f3f3f3f;
};

File: LintCode/python/Maximum Product Subarray.py
class Solution:
    # @param nums: an integer[]
    # @return: an integer
    def maxProduct(self, nums):
        if len(nums) == 1:
            return nums[0]
            
        maxi = 1
        mini = 1
        ans = -0xdeadbeef
        
        for cur in nums:
            a = maxi * cur
            b = mini * cur
            
            ans = max(ans, a, b, cur)
            
            maxi = max(cur, a, b)
            mini = min(cur, a, b)
        
        return ans
